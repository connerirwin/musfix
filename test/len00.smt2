; Polymorphic parameters are currently replaced with ints

; (qualif ListZ ((v [@0]) (>= (len v) 0)))
(qualif ListZ ((v [Int])) (>= (len v) 0))
; (qualif ListZ ((v [a])) (>= (len v) 0))

; uninterpreted function len that takes 2 polymorphic parameters,
; @(0) and @(1) and returns an Int

; number is the # of polymorphic params
; TODO why is it explicitly stated?
; constant len : (func(2, [(@(0)  @(1)); Int]))

; should the function type be separated?
; (uninterp len (@0 @1) Int)
(uninterp len (Int Int) Int)

; TODO add support for type constructors
; (wf [$k0] ((v [(Tuple Int a)]))))
(wf [$k0] ((v [Silly Int])))

(constraint
(forall ((v [(Tuple Int a)]))
;  (forall ((v [(Silly Int)]))
          (=> (>= (len v) 0) ([$k0]))))

(constraint
; (forall ((v [(Tuple Int a)]))
  (forall ((v [(Silly Int)]))
          (=> ([$k0]) (>= (len v) 0))))
