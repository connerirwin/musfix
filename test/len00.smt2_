(qualif ListZ ((v (Set @0))) (>= (len v) 0))

(define-fun len (@0 @1) Int)

(wf (Set $k0) ((v (Set (Tuple Int @a)))))

(constraint
  (forall ((v (Set (Tuple Int @a))))
    (=> (>= (len v) 0)
        ((Set $k0)))))

(constraint
  (forall ((v (Set (Tuple Int @a))))
    (=> ((Set $k0))
        (>= (len v) 0))))
