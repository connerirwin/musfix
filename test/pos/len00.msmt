; Uninterpreted Sorts
(declare-sort List 1)
(declare-sort Tuple 2) ; manual
(declare-sort Obj 0) ; manual

; Constants


; Distinct Constants


; Uninterpreted Functions
(declare-fun len ((@a0 @a1)) Int)


; Qualifiers
(qualif ListZ ((v (List @a0))) (>= (len v) 0))


; Well-formedness Constraints
(wf $k0 ((v (List (Tuple Int Obj)))))


; Horn Constraints
(constraint (forall ((yoink (List (Tuple Int Obj)))(v1 (List (Tuple Int Obj)))) (=> (and (>= (len yoink) 0) (= v1 yoink)) ($k0 v1))))
(constraint (forall ((v2 (List (Tuple Int Obj)))) (=> ($k0 v2) (>= (len v2) 0))))
