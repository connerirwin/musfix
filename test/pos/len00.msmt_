; Uninterpreted Sorts
(declare-sort List 1)


; Constants


; Distinct Constants


; Uninterpreted Functions
(declare-fun len ((@a0 @a1)) Int)


; Qualifiers
(qualif ListZ ((v (List @a0))) (>= (len v) 0))


; Well-formedness Constraints
(wf $k0 ((v (List (Tuple Int Obj a)))))


; Horn Constraints
(constraint (forall ((yoink (List (Tuple Int Obj a)))(v1 (List (Tuple Int Obj a)))) (=> (and (>= (apply len yoink) 0) (= v1 yoink)) ($k0 v1))))
(constraint (forall ((v2 (List (Tuple Int Obj a)))) (=> ($k0 v2) (>= (apply len v2) 0))))
