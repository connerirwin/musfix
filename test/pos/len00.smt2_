; Uninterpreted Sorts


; Uninterpreted Functions
(declare-fun len ((@a0 @a1)) Int)


; Qualifiers
(qualif ListZ ((v ([] @a0))) (>= (len v) 0))


; Well-formedness constraints
(wf $k0 ((v ([] (Tuple Int [a])))))


; Horn constraints
(constraint (forall ((yoink ([] (Tuple Int [a])))(v1 ([] (Tuple Int [a])))) (=> (and (>= (apply len yoink) 0) (= v1 yoink)) ($k0 v1))))
(constraint (forall ((v2 ([] (Tuple Int [a])))) (=> ($k0 v2) (>= (apply len v2) 0))))
