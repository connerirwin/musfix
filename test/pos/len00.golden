
Preparing to find greatest fixpoint...

Inputs
--------
Reading from file test/pos/len00.msmt
SortDecl "List" 1
SortDecl "Tuple" 2
SortDecl "Obj" 0
UninterpFunction "len" [DataS "_any" [VarS "a1"]] IntS
Qualifier "ListZ" [Var (DataS "List" [VarS "a0"]) "v"] (Binary Ge (Func IntS "len" [Var (DataS "List" [VarS "a1"]) "v"]) (IntLit 0))
WFConstraint "$k0" [Var (DataS "List" [DataS "Tuple" [IntS,DataS "Obj" []]]) "v"]
HornConstraint [Var (DataS "List" [DataS "Tuple" [IntS,DataS "Obj" []]]) "yoink",Var (DataS "List" [DataS "Tuple" [IntS,DataS "Obj" []]]) "v1"] (Binary Implies (Binary And (Binary Ge (Func IntS "len" [Var (DataS "List" [DataS "Tuple" [IntS,DataS "Obj" []]]) "yoink"]) (IntLit 0)) (Binary Eq (Var (DataS "List" [DataS "Tuple" [IntS,DataS "Obj" []]]) "v1") (Var (DataS "List" [DataS "Tuple" [IntS,DataS "Obj" []]]) "yoink"))) (Unknown (fromList [("v",Var (DataS "List" [DataS "Tuple" [IntS,DataS "Obj" []]]) "v1")]) "$k0"))
HornConstraint [Var (DataS "List" [DataS "Tuple" [IntS,DataS "Obj" []]]) "v2"] (Binary Implies (Unknown (fromList [("v",Var (DataS "List" [DataS "Tuple" [IntS,DataS "Obj" []]]) "v2")]) "$k0") (Binary Ge (Func IntS "len" [Var (DataS "List" [DataS "Tuple" [IntS,DataS "Obj" []]]) "v2"]) (IntLit 0)))

QMAP
--------
fromList [("$k0",QSpace {_qualifiers = [], _maxCount = 0})]

Candidates
--------
[]


Final candidates: 

